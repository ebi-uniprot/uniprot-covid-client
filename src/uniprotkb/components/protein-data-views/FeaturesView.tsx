import React, { Fragment, useCallback } from 'react';
import { html, TemplateResult } from 'lit-html';
import ProtvistaTrack from 'protvista-track';
import ProtvistaManager from 'protvista-manager';
import ProtvistaSequence from 'protvista-sequence';
import ProtvistaNavigation from 'protvista-navigation';
import { v1 } from 'uuid';
import { loadWebComponent } from '../../../shared/utils/utils';
import { Evidence } from '../../types/modelTypes';
import FeatureType from '../../types/featureType';
import { UniProtProtvistaEvidenceTag } from './UniProtKBEvidenceTag';
import FeaturesTableView, { FeaturesTableCallback } from './FeaturesTableView';
import { Xref } from '../../types/commentTypes';

enum LocationModifier {
  EXACT = 'EXACT',
  OUTSIDE = 'OUTSIDE',
  UNSURE = 'UNSURE',
  UNKNOWN = 'UNKNOWN',
}

type FeatureLocation = {
  value: number;
  modifier: LocationModifier;
};

export type FeatureData = {
  type: FeatureType;
  featureId?: string;
  description?: string;
  location: {
    start: FeatureLocation;
    end: FeatureLocation;
    sequence?: string;
  };
  alternativeSequence?: {
    originalSequence?: string;
    alternativeSequences: string[];
  };
  evidences?: Evidence[];
  featureCrossReference?: Xref;
}[];

export type ProtvistaFeature = {
  accession?: string;
  type: string;
  description: string;
  evidences: Evidence[];
  start: number;
  end: number;
  startModifier: LocationModifier;
  endModifier: LocationModifier;
};

type FeatureProps = {
  sequence?: string;
  features: FeatureData;
};

// TODO: looks like this could be merged with ProtvistaFeature?
export type ProcessedFeature = {
  protvistaFeatureId?: string;
  featureId?: string;
  start: number;
  end: number;
  startModifier: LocationModifier;
  endModifier: LocationModifier;
  type: FeatureType;
  description?: string;
  evidences?: Evidence[];
  sequence?: string;
};

const processData = (
  data: FeatureData,
  sequence?: string
): ProcessedFeature[] =>
  data.map(
    (feature): ProcessedFeature => ({
      protvistaFeatureId: feature.featureId || v1(),
      featureId: feature.featureId,
      start: feature.location.start.value,
      end: feature.location.end.value,
      startModifier: feature.location.start.modifier,
      endModifier: feature.location.end.modifier,
      type: feature.type,
      description: feature.description,
      evidences: feature.evidences,
      sequence: sequence?.substring(
        feature.location.start.value - 1,
        feature.location.end.value
      ),
    })
  );

const FeaturesView: React.FC<FeatureProps> = ({
  sequence,
  features,
}): JSX.Element | null => {
  loadWebComponent('protvista-track', ProtvistaTrack);
  loadWebComponent('protvista-manager', ProtvistaManager);
  loadWebComponent('protvista-sequence', ProtvistaSequence);
  loadWebComponent('protvista-navigation', ProtvistaNavigation);

  const processedData = processData(features, sequence);

  const getColumnConfig = (evidenceTagCallback: FeaturesTableCallback) => ({
    type: {
      label: 'Type',
      resolver: (d: ProcessedFeature): string => d.type,
    },
    id: {
      label: 'ID',
      resolver: (d: ProcessedFeature): string => d.featureId || '',
    },
    positions: {
      label: 'Positions',
      resolver: (d: ProcessedFeature): string =>
        `${d.startModifier === LocationModifier.UNKNOWN ? '?' : d.start}-${
          d.endModifier === LocationModifier.UNKNOWN ? '?' : d.end
        }`,
    },
    description: {
      label: 'Description',
      resolver: (d: ProcessedFeature): TemplateResult =>
        html`
          ${d.description}
          ${d.evidences &&
          UniProtProtvistaEvidenceTag(d.evidences, evidenceTagCallback)}
        `,
    },
    featureId: {
      label: 'ID',
      resolver: (d: ProtvistaFeature) => (d.accession ? d.accession : '-'),
    },
    sequence: {
      label: 'Sequence',
      child: true,
      resolver: (d: ProcessedFeature) => d?.sequence || '',
    },
  });

  const setTrackData = useCallback(
    (node): void => {
      if (node) {
        // eslint-disable-next-line no-param-reassign
        node.data = processedData;
      }
    },
    [processedData]
  );

  if (processedData.length === 0) {
    return null;
  }

  return (
    <Fragment>
      <h3>Features</h3>
      <protvista-manager attributes="highlight displaystart displayend selectedid">
        {sequence && (
          <Fragment>
            <protvista-navigation length={sequence.length} />
            <protvista-track
              ref={setTrackData}
              length={sequence.length}
              layout="non-overlapping"
            />
            <protvista-sequence
              sequence={sequence}
              length={sequence.length}
              height="20"
            />
          </Fragment>
        )}
        <FeaturesTableView
          data={processedData}
          getColumnConfig={getColumnConfig}
        />
      </protvista-manager>
    </Fragment>
  );
};

export default FeaturesView;
