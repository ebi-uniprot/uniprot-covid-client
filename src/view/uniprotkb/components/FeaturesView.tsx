import React, { Fragment, useCallback } from 'react';
import { html, TemplateResult } from 'lit-html';
import ProtvistaTrack from 'protvista-track';
import ProtvistaManager from 'protvista-manager';
import ProtvistaSequence from 'protvista-sequence';
import ProtvistaNavigation from 'protvista-navigation';
import { loadWebComponent } from '../../../utils/utils';
import { Evidence } from '../../../model/types/modelTypes';
import FeatureType from '../../../model/types/FeatureType';
import { UniProtProtvistaEvidenceTag } from '../../../components/UniProtEvidenceTag';
import FeaturesTableView, { FeaturesTableCallback } from './FeaturesTableView';
import { Xref } from '../../../model/types/CommentTypes';

enum LocationModifier {
  EXACT = 'EXACT',
  OUTSIDE = 'OUTSIDE',
  UNSURE = 'UNSURE',
  UNKNOWN = 'UNKNOWN',
}

type FeatureLocation = {
  value: number;
  modifier: LocationModifier;
};

export type FeatureData = {
  type: FeatureType;
  featureId?: string;
  description?: string;
  location: {
    start: FeatureLocation;
    end: FeatureLocation;
    sequence?: string;
  };
  alternativeSequence?: {
    originalSequence?: string;
    alternativeSequences: string[];
  };
  evidences?: Evidence[];
  featureCrossReference?: Xref;
}[];

export type ProtvistaFeature = {
  type: string;
  description: string;
  evidences: Evidence[];
  start: number;
  end: number;
  startModifier: LocationModifier;
  endModifier: LocationModifier;
};

type FeatureProps = {
  sequence?: string;
  features: FeatureData;
};

export type ProcessedFeature = {
  accession: string | undefined;
  start: number;
  end: number;
  startModifier: LocationModifier;
  endModifier: LocationModifier;
  type: FeatureType;
  description: string | undefined;
  evidences: Evidence[] | undefined;
};

const processData = (data: FeatureData): ProcessedFeature[] =>
  data.map(
    (feature): ProcessedFeature => ({
      accession: feature.featureId,
      start: feature.location.start.value,
      end: feature.location.end.value,
      startModifier: feature.location.start.modifier,
      endModifier: feature.location.end.modifier,
      type: feature.type,
      description: feature.description,
      evidences: feature.evidences,
    })
  );

const FeaturesView: React.FC<FeatureProps> = ({
  sequence,
  features,
}): JSX.Element | null => {
  loadWebComponent('protvista-track', ProtvistaTrack);
  loadWebComponent('protvista-manager', ProtvistaManager);
  loadWebComponent('protvista-sequence', ProtvistaSequence);
  loadWebComponent('protvista-navigation', ProtvistaNavigation);

  const processedData = processData(features);

  const getColumnConfig = (evidenceTagCallback: FeaturesTableCallback) => ({
    type: {
      label: 'Type',
      resolver: (d: ProtvistaFeature): string => d.type,
    },
    positions: {
      label: 'Positions',
      resolver: (d: ProtvistaFeature): string =>
        `${d.startModifier === LocationModifier.UNKNOWN ? '?' : d.start}-${
          d.endModifier === LocationModifier.UNKNOWN ? '?' : d.end
        }`,
    },
    description: {
      label: 'Description',
      resolver: (d: ProtvistaFeature): TemplateResult =>
        html`
          ${d.description}
          ${d.evidences &&
            UniProtProtvistaEvidenceTag(d.evidences, evidenceTagCallback)}
        `,
    },
  });

  const setTrackData = useCallback(
    (node): void => {
      if (node) {
        // eslint-disable-next-line no-param-reassign
        node.data = processedData;
      }
    },
    [processedData]
  );

  if (processedData.length === 0) {
    return null;
  }

  return (
    <Fragment>
      <h3>Features</h3>
      <protvista-manager attributes="highlight displaystart displayend">
        {sequence && (
          <Fragment>
            <protvista-navigation length={sequence.length} />
            <protvista-track
              ref={setTrackData}
              length={sequence.length}
              layout="non-overlapping"
            />
            <protvista-sequence
              sequence={sequence}
              length={sequence.length}
              height="20"
            />
          </Fragment>
        )}
        <FeaturesTableView
          data={processedData}
          getColumnConfig={getColumnConfig}
        />
      </protvista-manager>
    </Fragment>
  );
};

export default FeaturesView;
