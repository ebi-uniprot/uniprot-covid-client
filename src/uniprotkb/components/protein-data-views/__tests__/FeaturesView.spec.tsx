import React from 'react';
import { render, cleanup } from '@testing-library/react';
import FeaturesView from '../FeaturesView';
import FeaturesUIDataJson from './__mocks__/featuresUIData.json';

afterEach(cleanup);

describe('FeaturesView component', () => {
  test('it renders without crashing', () => {
    const { asFragment } = render(
      <FeaturesView
        features={FeaturesUIDataJson}
        sequence="ASDASDASASDASDASDSASD"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
