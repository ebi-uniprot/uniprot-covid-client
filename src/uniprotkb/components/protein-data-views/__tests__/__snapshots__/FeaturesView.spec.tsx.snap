// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeaturesView component it renders without crashing 1`] = `
<DocumentFragment>
  <h3>
    Features
  </h3>
  <protvista-manager
    attributes="highlight displaystart displayend"
  >
    <protvista-navigation
      length="21"
    />
    <protvista-track
      layout="non-overlapping"
      length="21"
    />
    <protvista-sequence
      height="20"
      length="21"
      sequence="ASDASDASASDASDASDSASD"
    />
    <protvista-datatable />
    <div
      class="evidence-tag-content false"
    />
  </protvista-manager>
</DocumentFragment>
`;
